File: concepts/03_errors_and_resources/02_exception_propagation_and_matching/lua/test_02_rethrow_and_completion_precedence.lua
-- Common question: what happens when cleanup and the body both fail?
-- Inputs: a body error, __close inspection, a closing error, and explicit rethrow.
-- Observations: current error passed to cleanup, replacement precedence, and original identity.
-- polyglot-family: errors_and_resources
-- polyglot-concept: exception_propagation_and_matching
-- polyglot-related: languages/lua/language/07_errors_and_resources/test_054_close_failures_and_pending_cleanup.lua

local t = require("support.assertions")
local original = {kind = "body"}
local seen
local resource = setmetatable({}, {
    __close = function(_, current)
        seen = current
        error({kind = "close", cause = current}, 0)
    end,
})

local ok, final = pcall(function()
    local handle <close> = resource
    t.same(handle, resource)
    error(original)
end)
t.falsey(ok)
t.same(seen, original)
t.equal(final.kind, "close")
t.same(final.cause, original)

local rethrow_ok, rethrown = pcall(function()
    local inner_ok, inner_error = pcall(function() error(original) end)
    if not inner_ok then error(inner_error) end
end)
t.falsey(rethrow_ok)
t.same(rethrown, original)

t.done()

File: concepts/03_errors_and_resources/02_exception_propagation_and_matching/ruby/test_02_rethrow_and_completion_precedence.rb
# 共同问题：重新抛出是否保留异常身份，body 与 ensure 都失败时哪个完成占优。
# 输入：bare raise、body 异常和 ensure 异常；观察：身份保留、最终异常及自动 cause。
# polyglot-family: errors_and_resources
# polyglot-concept: exception_propagation_and_matching
# polyglot-related: languages/ruby/language/08_exceptions_resources_and_loading/
# polyglot-related+: test_058_exception_cause_and_backtrace.rb

require "assertions"

A = PolyglotAssertions

original = RuntimeError.new("body")
rethrown = begin
  begin
    raise original
  rescue
    raise
  end
rescue => error
  error
end
A.same(original, rethrown)

replacement = begin
  begin
    raise original
  ensure
    raise ArgumentError, "cleanup"
  end
rescue => error
  error
end
A.same(ArgumentError, replacement.class)
A.equal("cleanup", replacement.message)
A.same(original, replacement.cause)

A.done
